feat: add a built-in changelog page (updated)

UltimatChamp opened the pull request:
Closes #43.
Closes #50.

## Changes:

### Light Mode

![Screenshot 2024-12-13 122736](https://github.com/user-attachments/assets/5a0ca7e4-56c6-4ad9-bb26-b15a637ef66b)

### Dark Mode

![Screenshot 2024-12-13 122812](https://github.com/user-attachments/assets/ebde4453-c549-41df-b8fe-5ddbb9b5ac29)

https://github.com/user-attachments/assets/4a849d63-6547-4aaf-8fc4-69f6aec83c6a

---

- The changelog can also be accessed via: http://neoforged.net/changelog

- The changelog can also be found here:
  <img src="https://github.com/user-attachments/assets/641ab17b-60ca-43f1-830f-c29c33dfc703" width="200px">

- The changelog link in the _Installer Files_ section has also been replaced.

- Added the ability to install a specific version from the changelog.

- Fix compatibility with the latest version of Hugo.

---

<details><summary>Old</summary>

- A _latest_ and a _legacy_(1.20.1) changelog pages.
  <img src="https://github.com/user-attachments/assets/40f7724c-e3b0-4837-b363-0482e3b98ac4" width="400px"><br>
  <img src="https://github.com/user-attachments/assets/81e81b5b-6260-46ad-861c-a7852e22ea70" width="400px">
</details>

---

Preview URL: https://pr-49.neoforged-website-previews.pages.dev

UltimatChamp commented:
Hmm...
1.20.1 throws this error:
![image](https://github.com/user-attachments/assets/cfbaae45-f959-4f77-968e-a84ac05a4867)
![image](https://github.com/user-attachments/assets/83067846-7fda-4662-9cc1-d56fd74a2ef1)

UltimatChamp commented:
Marking this as draft, as I have some plans instead of just a code-block.

UltimatChamp commented:
Current progress:

![image](https://github.com/user-attachments/assets/7c3b5af0-8963-4d4e-9b39-22b91359c57d)

This currently lacks css and stuff. Btw, open to your suggestions.

UltimatChamp commented:
![image](https://github.com/user-attachments/assets/d0f39782-0984-46e2-80a2-2c97dd03072b)

more progess

UltimatChamp commented:
### More Changes

- Fix the changelog in light mode.
- Add the ability to install a specific version from the changelog.
- Make changelog version descriptions more distinguishable.
- Add support for issue links and code blocks in the changelog.
- Fix compatibility with the latest version of Hugo **(the Neoforge pages deployment app will also need to use the latest version of Hugo)**.

Technici4n commented:
Looks amazing!

- The links for each version download all point to the latest version.
- The changelogs category is a bit weird, there should be a way to move the page outside of any category?

UltimatChamp commented:
> * The links for each version download all point to the latest version.

Whoops! Fixed now.

> * The changelogs category is a bit weird, there should be a way to move the page outside of any category?

Yeah, this PR was made way back, when a legacy version (1.20.1) of Neoforge existed.
I've now removed the category, and the changelog can now be accessed here: http://neoforged.net/changelog

UltimatChamp commented:
> We can do this, but I was actually thinking about outputting a JSON changelog too to also link to each commit. This code is string-parsing the text based changelog, which isn't ideal.

That would be very convenient. The current code uses some quirks to figure things out, like ` - ` for new lines and `   ` for their descriptions.

Technici4n commented on themes/mainroad/layouts/partials/comments.html:
```diff
@@ -1,4 +1,4 @@
-{{ if and .Site.DisqusShortname (index .Params "comments" | default "true") (not .Site.IsServer) }}
+{{ if and .Site.Config.Services.Disqus.Shortname (index .Params "comments" | default "true") (not hugo.IsServer) }}
```
Why these changes?

UltimatChamp replied:
To prevent crashing with the latest Hugo.

Technici4n commented on layouts/shortcodes/changelog.html:
```diff
@@ -0,0 +1,17 @@
+{{ if not ( .Page.Scratch.Get "nfJS") }}
```
you sure about this scratch stuff? At least call it `changelogJS`.

UltimatChamp replied:
On doing some research, this line is required to ensure that each js file is included only once per page...? So I guess you're right.

Technici4n reviewed (changes requested):
I'm happy to accept this PR as-is, we can easily iterate on it later. Two questions though.

Technici4n commented on assets/js/changelog.js:
```diff
@@ -0,0 +1,86 @@
+                const installerUrl = `${DOWNLOAD_URL}/${gav}/${lineVersion}/${fn}-${lineVersion}-installer.jar`;
+                line = line.replace("`" + lineVersion + "`", `<a href="${installerUrl}" class="changelog_version"><code>${lineVersion}</code></a>`);
+
+                line = line.replace(" - ", `<li class="changelog_item" title="Install ${lineVersion} for Minecraft ${mcvers}">`);
```
You should remove the `title` from the list item because it adds this nonsensical tooltip:
![image](https://github.com/user-attachments/assets/19aa4058-f0df-49af-a2f9-c61a82c17f70)

It should only be added to the `a` element.

UltimatChamp replied:
Fixed!

Technici4n commented:
Thanks a lot, this is nice! Let's iterate from here.
